fiftyone/utils/youtubedl.py

```diff
@@ -0,0 +1,343 @@
+
+        clip_segments = [
+            (10, 25),
+            (11.1, 20.2),
+            None,
+            (None, 8.3),
+        ]
+
+    Args:
+        urls: either a list of video urls, or a dict
+            mapping these urls to locations on disk. If `urls` is a list, then
+            the `download_dir` argument is required
+        download_dir (None): the output directory used to store the downloaded
+            videos. This is only used if `urls` is a list
+        clip_segments (None): a list of int or float tuples indicating start
+            and end times in seconds for downloading segments of videos
+        max_videos (None): the maximum number of videos to download from the
+            given ``urls``. By default, all ``urls`` will be downloaded
+        num_workers (None): the number of processes to use when downloading
+            individual video. By default, ``multiprocessing.cpu_count()`` is
+            used
+        ext (None): the extension to use to store the downloaded videos or a
+            list of extensions corresponding to the given ``urls``. By default,
+            the default extension of each video is used 
+        buffer_seconds (5): the number of additional seconds before and after
+            each clip to download to ensure nearest keyframes are captured if
+            `clip_segments` is provided
+
+    Returns:
+        urls of successfully downloaded videos
+        a dict of download errors and the corresponding video urls
+    """
+    num_workers = _parse_num_workers(num_workers)
+    tasks = _build_tasks_list(
+        urls, clip_segments, download_dir, ext, buffer_seconds=buffer_seconds
+    )
+
+    if max_videos is None:
+        max_videos = len(tasks)
+
+    if num_workers == 1:
+        downloaded, errors = _single_thread_download(tasks, max_videos)
+
+    else:
+        downloaded, errors = _multi_thread_download(
+            tasks, max_videos, num_workers
+        )
+
+    return downloaded, errors
+
+
+def _parse_num_workers(num_workers):
+    if num_workers is None:
+        if os.name == "nt":
+            # Default to 1 worker for Windows OS
+            return 1
+        return multiprocessing.cpu_count()
+
+    if not isinstance(num_workers, int) or num_workers < 1:
+        raise ValueError(
+            "The `num_workers` argument must be a positive integer or `None` "
+            "found %s" % str(type(num_workers))
+        )
+    return num_workers
+
+
+def _build_tasks_list(
+    urls, clip_segments, download_dir, ext, buffer_seconds=5
+):
+    if isinstance(urls, str):
+        urls = [urls]
+    if isinstance(clip_segments, tuple):
+        clip_segments = [clip_segments]
+
+    if isinstance(urls, list):
+        if download_dir is None:
+            raise ValueError(
+                "When `urls` is a list, `download_dir` is required but was found to be `None`."
+            )
+        urls = {url: None for url in urls}
+
+        etau.ensure_dir(download_dir)
```

Comment:
Rather than `ensure_dir()` here, better to ensure the base directory of every output path (they could be arbitrary if the user provided them) at actual download time.